### PR TITLE
support init freegas

### PR DESCRIPTION
**Pull request by chengzhinei**
Closes #<issue number>.

### Reviewers

Main reviewers:

- @John
- @Doe

Codeowner reviewers:

- @-Alice
- @-Bob

### CODE REVIEW

**Review by ylsGit — APPROVED** [review]

File: pool/pool.go [hunk 1]
```diff
@@ -535,8 +536,20 @@ func (p *Pool) validateTx(ctx context.Context, poolTx Transaction) error {
...
 
-	// Reject transactions with a gas price lower than the minimum gas price
-	if !isFreeGasAddress(p.cfg.FreeGasAddress, from) || !poolTx.IsClaims { // XLayer handle
+	freeGp := false
+	if isFreeGasAddress(p.cfg.FreeGasAddress, from) && poolTx.IsClaims { // claim tx
+		freeGp = true
+	} else if currentNonce < getFreeGasCountPerAddr(p.cfg.FreeGasCountPerAddr) { // free-gas tx by count
+		freeGp, err = p.storage.IsFreeGasAddr(ctx, from)
+		if err != nil {
+			return err
```
Comment (giskook) [inline pool/pool.go:545]: Maybe we can wrap this err, so we can know, here's a unexpected behavior?

File: pool/pool.go [hunk 2]
```diff
@@ -604,6 +617,23 @@ func (p *Pool) validateTx(ctx context.Context, poolTx Transaction) error {
 		return err
 	}
 
+	if poolTx.IsClaims {
+		// check the address of to can be free-gas
+		inputHex := hex.EncodeToHex(poolTx.Data())
+		addrHex := "0x" + inputHex[4514:4554]
```
Comment (giskook) [inline pool/pool.go:623]: I think we should decode the poolTx.Data here with ABI? It's not clear what's the inputHex[4514:4554] is.

File: pool/pool.go [hunk 3]
```diff
@@ -604,6 +617,23 @@ func (p *Pool) validateTx(ctx context.Context, poolTx Transaction) error {
...
+		addr := common.HexToAddress(addrHex)
+
+		nonce, err := p.state.GetNonce(ctx, addr, lastL2Block.Root())
+		if err != nil {
+			log.Errorf("failed to getcount while check free address", err)
+		}
+		if nonce < getFreeGasCountPerAddr(p.cfg.FreeGasCountPerAddr) {
+			if err = p.storage.AddFreeGasAddr(ctx, addr); err != nil {
+				return err
```
Comment (giskook) [inline pool/pool.go:632]: Should we wrap the err as well?

File: sequencer/sequencer.go [hunk 4]
```diff
@@ -233,20 +233,36 @@ func (s *Sequencer) loadFromPool(ctx context.Context) {
 	}
 }
 
+func (s *Sequencer) checkFreeGas(tx pool.Transaction, txTracker *TxTracker) (freeGp, claimTx bool) {
```
Comment (zjg555543) [inline sequencer/sequencer.go:236]: I think this code can be added to the sequencer_xlayer.go file to avoid file conflicts

File: pool/pool.go [hunk 5]
```diff
@@ -604,6 +647,37 @@ func (p *Pool) validateTx(ctx context.Context, poolTx Transaction) error {
 		return err
 	}
 
+	if getEnableFreeGasByNonce(p.cfg.EnableFreeGasByNonce) {
```
Comment (zjg555543) [inline pool/pool.go:650]: Move code into the xxxx_xlayer.go file to avoid file conflicts

File: pool/pool.go [hunk 6]
```diff
@@ -177,6 +182,18 @@ func (p *Pool) StartPollingMinSuggestedGasPrice(ctx context.Context) {
 	}()
 }
 
+// AddDynamicGp cache the dynamic gas price of L2
+func (p *Pool) AddDynamicGp(dgp *big.Int) {
```
Comment (zjg555543) [inline pool/pool.go:186]: Move code into the xxxx_xlayer.go file to avoid file conflicts

File: pool/pool.go [hunk 7]
```diff
@@ -402,6 +419,18 @@ func (p *Pool) GetGasPrices(ctx context.Context) (GasPrices, error) {
 	return GasPrices{L1GasPrice: l1GasPrice, L2GasPrice: l2GasPrice}, err
 }
 
+// GetDynamicGasPrice returns the current L2 dynamic gas price
+func (p *Pool) GetDynamicGasPrice() *big.Int {
+	p.dgpMux.RLock()
```
Comment (zjg555543) [inline pool/pool.go:424]: Move code into the xxxx_xlayer.go file to avoid file conflicts

File: pool/pool.go [hunk 8]
```diff
@@ -604,6 +647,37 @@ func (p *Pool) validateTx(ctx context.Context, poolTx Transaction) error {
...
+		inputHex := hex.EncodeToHex(poolTx.Data())
+		// hard code
+		if isFreeGasExAddress(p.cfg.FreeGasExAddress, from) &&
+			strings.HasPrefix(inputHex, ExWithdrawalMethodSignature) &&
+			len(inputHex) > 74 { // erc20 contract transfer
+			addrHex := "0x" + inputHex[10:74]
+			freeGpAddr = common.HexToAddress(addrHex)
+		} else if poolTx.IsClaims &&
+			len(inputHex) > 4554 { // bridge contract claim
+			addrHex := "0x" + inputHex[4490:4554]
```
Comment (zjg555543) [inline pool/pool.go:662]: These magic nums a little diffcult to undersand.

File: pool/pool.go [hunk 9]
```diff
@@ -535,8 +564,22 @@ func (p *Pool) validateTx(ctx context.Context, poolTx Transaction) error {
 		}
 	}
 
-	// Reject transactions with a gas price lower than the minimum gas price
-	if !isFreeGasAddress(p.cfg.FreeGasAddress, from) || !poolTx.IsClaims { // XLayer handle
+	freeGp := false
+	if isFreeGasAddress(p.cfg.FreeGasAddress, from) && poolTx.IsClaims { // claim tx
+		freeGp = true
+	} else if getEnableFreeGasByNonce(p.cfg.EnableFreeGasByNonce) &&
+		currentNonce < getFreeGasCountPerAddr(p.cfg.FreeGasCountPerAddr) { // free-gas tx by count
```
Comment (zjg555543) [inline pool/pool.go:571]: Should it use poolTx.Nonce()?

File: jsonrpc/endpoints_eth.go [hunk 10]
```diff
@@ -1043,6 +1043,9 @@ func (e *EthEndpoints) tryToAddTxToPool(input, ip string) (interface{}, types.Er
 		return RPCErrorResponse(types.InvalidParamsErrorCode, "invalid tx input", err, false)
 	}
 	log.Infof("adding TX to the pool: %v", tx.Hash().Hex())
+
+	dgp := getDynamicGp(e.cfg.DynamicGP.Enabled, e.dgpMan.lastPrice)
```
Comment (giskook) [inline jsonrpc/endpoints_eth.go:1047]: e.cfg.DynamicGP.Enabled may changed by apollo, so maybe we should get enabled wih apollo's way?

File: pool/pool_xlayer.go [hunk 11]
```diff
@@ -84,3 +90,88 @@ func (p *Pool) GetMinSuggestedGasPriceWithDelta(ctx context.Context, delta time.
...
+		if isFreeAddr && poolTx.Nonce() < freeGasCountPerAddrConfig {
+			if poolTx.Gas() > getFreeGasLimit(p.cfg.FreeGasLimit) {
+				return false, fmt.Errorf("gas-free transaction with too high gas limit")
+			}
+			return true, nil
+		} else if isFreeAddr && poolTx.Nonce() >= freeGasCountPerAddrConfig {
+			return false, fmt.Errorf("you are no longer eligible for gas-free transactions because for each new address, only the first %d transactions(address nonce less than %d) can be gas-free",
+				freeGasCountPerAddrConfig,
+				freeGasCountPerAddrConfig)
+		} else if !isFreeAddr {
```
Comment (giskook) [inline pool/pool_xlayer.go:126]: Maybe `else` equal to `else if`, and I think with `else` is better.

File: pool/pool_xlayer.go [hunk 12]
```diff
@@ -84,3 +90,88 @@ func (p *Pool) GetMinSuggestedGasPriceWithDelta(ctx context.Context, delta time.
...
+		_, l2Gp := p.GetL1AndL2GasPrice()
+		dgp = new(big.Int).SetUint64(l2Gp)
+	}
+	return dgp
+}
+
+func (p *Pool) checkFreeGp(ctx context.Context, poolTx Transaction, from common.Address) (bool, error) {
+	if isFreeGasAddress(p.cfg.FreeGasAddress, from) && poolTx.IsClaims { // claim tx
+		return true, nil
+	} else if getEnableFreeGasByNonce(p.cfg.EnableFreeGasByNonce) && poolTx.GasPrice().Cmp(big.NewInt(0)) == 0 { // free-gas tx by count
```
Comment (giskook) [inline pool/pool_xlayer.go:109]: Maybe instead of `if` with `else if`?

File: pool/pool.go (hunk 8 above)
Comment (ylsGit) [reply pool/pool.go]: Optimize later

**Review by giskook — APPROVED** [review]

**Review by zjg555543 — APPROVED** [review]

**Pull request merged**